common/src/main/java/io/netty/util/Recycler.java

```diff
@@ -189,558 +175,100 @@ public final boolean recycle(T o, Handle<T> handle) {
             return false;
         }
 
-        DefaultHandle<T> h = (DefaultHandle<T>) handle;
-        if (h.stack.parent != this) {
-            return false;
-        }
-
-        h.recycle(o);
+        handle.recycle(o);
         return true;
     }
 
-    final int threadLocalCapacity() {
-        return threadLocal.get().elements.length;
-    }
-
     final int threadLocalSize() {
-        return threadLocal.get().size;
+        return threadLocal.get().pooledHandles.size();
     }
 
     protected abstract T newObject(Handle<T> handle);
 
+    @SuppressWarnings("ClassNameSameAsAncestorName") // Can't change this due to compatibility.
     public interface Handle<T> extends ObjectPool.Handle<T>  { }
 
-    @SuppressWarnings("unchecked")
     private static final class DefaultHandle<T> implements Handle<T> {
-        private static final AtomicIntegerFieldUpdater<DefaultHandle<?>> LAST_RECYCLED_ID_UPDATER;
+        private static final int STATE_CLAIMED = 0;
+        private static final int STATE_AVAILABLE = 1;
+        private static final AtomicIntegerFieldUpdater<DefaultHandle<?>> STATE_UPDATER;
         static {
-            AtomicIntegerFieldUpdater<?> updater = AtomicIntegerFieldUpdater.newUpdater(
-                    DefaultHandle.class, "lastRecycledId");
-            LAST_RECYCLED_ID_UPDATER = (AtomicIntegerFieldUpdater<DefaultHandle<?>>) updater;
+            AtomicIntegerFieldUpdater<?> updater = AtomicIntegerFieldUpdater.newUpdater(DefaultHandle.class, "state");
+            //noinspection unchecked
+            STATE_UPDATER = (AtomicIntegerFieldUpdater<DefaultHandle<?>>) updater;
```

Comment:
nit: why not just create this directly ? I think there is really no need for the extra static block here.

Reply:
Because that would require two casts in a row in order to get the generics to line up. And that would look ugly. We've done static blocks for this situation in other places, so it's not a new pattern.